File: code/java/684.java
class UF{
    private int count;
    private int[] parent;

    public int getCount(){
        return count;
    }

    public UF(int num){
        count = num;
        parent = new int[num];
        for(int i=0;i<num;i++) parent[i] = i;
    }

    public int find(int x){
        while(x!=parent[x]){
            parent[x] = parent[parent[x]];
            x = parent[x];
        }
        return x;
    }

    public boolean union(int p,int q){
        int rootP = find(p);
        int rootQ = find(q);
        if(rootP==rootQ) return true;
        parent[rootQ] = rootP;
        count-=1;
        return false;
    }
}

class Solution {
    public int[] findRedundantConnection(int[][] edges) {
        UF uf = new UF(edges.length);
        for(int[] edge:edges){
            if(uf.union(edge[0]-1,edge[1]-1)) return edge;
        }
        return null;
    }
}

File: code/python/684.py
class UF:
    def __init__(self, count):
        self.count = count
        self.parent = [i for i in range(count)]

    def find(self, x):
        while self.parent[x] != x:
            self.parent[x] = self.parent[self.parent[x]]
            x = self.parent[x]
        return x

    def union(self, p, q):
        rootP = self.find(p)
        rootQ = self.find(q)
        # 该边能够构成环
        if rootP == rootQ:
            return False
        self.parent[rootP] = rootQ
        self.count -= 1
        return True

    def divide(self, p, q):
        pass


class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        uf = UF(len(edges))
        for edge in edges:
            x, y = edge
            if uf.union(x - 1, y - 1) == False:
                return edge